packages/@react-spectrum/table/src/TableView.tsx

```diff
@@ -983,31 +1162,94 @@ function TableRow({item, children, hasActions, ...otherProps}) {
     }
   }
 
+  let draggableItem: DraggableItemResult;
+  if (isTableDraggable) {
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    draggableItem = dragAndDropHooks.useDraggableItem({key: item.key, hasDragButton: true}, dragState);
+    if (isDisabled) {
+      draggableItem = null;
+    }
+  }
+  let droppableItem: DroppableItemResult;
+  let isDropTarget: boolean;
+  let dropIndicator: DropIndicatorAria;
+  let dropIndicatorRef = useRef();
+  if (isTableDroppable) {
+    let target = {type: 'item', key: item.key, dropPosition: 'on'} as DropTarget;
+    isDropTarget = dropState.isDropTarget(target);
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    dropIndicator = dragAndDropHooks.useDropIndicator({target}, dropState, dropIndicatorRef);
+  }
+
+  let dragButtonRef = React.useRef();
+  let {buttonProps: dragButtonProps} = useButton({
+    ...draggableItem?.dragButtonProps,
+    elementType: 'div'
+  }, dragButtonRef);
+
+  let props = mergeProps(
+    rowProps,
+    otherProps,
+    focusWithinProps,
+    focusProps,
+    hoverProps,
+    pressProps,
+    draggableItem?.dragProps,
+    // Remove tab index from list row if performing a screenreader drag. This prevents TalkBack from focusing the row,
+    // allowing for single swipe navigation between row drop indicator
+    dragAndDropHooks?.isVirtualDragging() && {tabIndex: null}
+  );
+
+  let dropProps = isDroppable ? droppableItem?.dropProps : {'aria-hidden': droppableItem?.dropProps['aria-hidden']};
+  let {visuallyHiddenProps} = useVisuallyHidden();
+
   return (
-    <div
-      {...props}
-      ref={ref}
-      className={
-        classNames(
-          styles,
-          'spectrum-Table-row',
-          {
-            'is-active': isPressed,
-            'is-selected': isSelected,
-            'spectrum-Table-row--highlightSelection': state.selectionManager.selectionBehavior === 'replace',
-            'is-next-selected': state.selectionManager.isSelected(item.nextKey),
-            'is-focused': isFocusVisibleWithin,
-            'focus-ring': isFocusVisible,
-            'is-hovered': isHovered,
-            'is-disabled': isDisabled,
-            'spectrum-Table-row--firstRow': isFirstRow,
-            'spectrum-Table-row--lastRow': isLastRow,
-            'spectrum-Table-row--isFlushBottom': isFlushWithContainerBottom
-          }
-        )
-      }>
-      {children}
-    </div>
+    <TableRowContext.Provider value={{dragButtonProps, dragButtonRef, isFocusVisibleWithin}}>
+      {isTableDroppable && isFirstRow &&
+        <InsertionIndicator
+          rowProps={props}
+          key={`${item.key}-before`}
+          target={{key: item.key, type: 'item', dropPosition: 'before'}} />
+      }
+      {isTableDroppable && !dropIndicator?.isHidden &&
+        <div role="button" {...visuallyHiddenProps} {...dropIndicator?.dropIndicatorProps} ref={dropIndicatorRef} />
```

Comment:
FYI I think this can't be directly inside a grid/rowgroup, it'll need to be inside a gridcell within a row.